NF: Migrate from `org.json.****` to jackson treemodel

TarekkMA opened the pull request:
## Pull Request template

## Purpose / Description
Use jackson tree model API instead of the android one. This will provide double the performance for large json objects.

## Fixes
Related to #8668

## Approach
Maintain `JSONObject`/`JSONArray` API but delegating all calls to `ObjectNode`/`ArrayNode`.

Objects like Model, DeckConfig, and others will inheart from `ObjectNode` and we will refactor them slowly to be their own independent models.

## How Has This Been Tested?

All existing tests work fine

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -153,199 +105,202 @@ public JSONObject(JSONObject copyFrom) {
-            return org.json.JSONObject.numberToString(number);
-        } catch (org.json.JSONException e) {
-            throw new RuntimeException (e);
+    public JSONArray names() {
```
This is a change in behavior. `names` used to return `null` if the list is empty. You may want to keep exact compatibility for the sake of safety.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -153,199 +105,202 @@ public JSONObject(JSONObject copyFrom) {
-        } catch (org.json.JSONException e) {
-            throw new RuntimeException (e);
+    public JSONArray names() {
+        Iterator<String> stringIterator = mNode.fieldNames();
```
Might be nice to do `return new JSONArray(keys())`, if `JSONArray` would accept an iterator of object as an extra constructor. I believe code would be easier to read and that may prove useful

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -153,199 +105,202 @@ public JSONObject(JSONObject copyFrom) {
 
+
     @CheckResult
     public JSONArray optJSONArray(String name) {
```
Since we don't have access to the javadoc of the overriden function anymore, I think it would be nice to copy it at least once for each kind of function. In this case, confirm that "return null if the key do not exists or if the associated value is not a json array"

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -363,18 +318,114 @@ public JSONObject deepClone() {
             }
             else if (get(key) instanceof JSONArray) {
                 clone.put(key, getJSONArray(key).deepClone());
+            } else if (get(key) instanceof JsonNode) {
```
`deepCloneInto` was necessary because `Model` (and similar classes) was inheriting `JSONObject`. If you decide to delegate instead of inheriting, maybe we don't need it anymore, since you can just create a new `Model` and uses `deepClone()` on its `JSONObject`. It was really useful because I wanted a JSONObject constructor returning a Model. This may change

TarekkMA replied:
okay I will remove it

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Deck.java:
```diff
@@ -25,6 +29,17 @@ public Deck(JSONObject json) {
+    }
+
+    @JsonValue
+    @Override
```
Can you please explain the point of overriding. It seems useless here since you just call super.
Does `@JsonValue` alone do something? In this case why not to put it on the overridden method too?

TarekkMA replied:
If we wish to use data binding API with these classes we should specify how to deserialize the object with `@ JsonCreator` and what is the value of the class when being serialized with `@JsonValue`

```java
// example of data binding API
final Deck deck = new ObjectMapper().readValue(source, Deck.class);
```

The super is already annotated and have tests (look `testDatabindingSerializeDeserialize` in both Array and Object test)

I don't know if adding annotating the super is enough, but I wanted to include it so reading the class alone would be enough to know how Jackson will de/serialize it.

Arthur-Milchior replied:
Thanks for the explanation. I'd appreciate to get it in javadoc, as future reader may not know jackson either

TarekkMA replied:
Alright, should it be on JSONObject or on all subclasses (Model, Deck, etc)?

TarekkMA replied:
I decided not to override `getRootJsonNode` and Jackson can infer it from the super class. 

Added java docs on `JSONObject/Array` about the use of the annotation for subclasses

Arthur-Milchior replied:
I'm fine with not repeating javadoc if it is exactly the same when you override.  I think I remember android studio being wise enough to select the last javadoc available in the hierarchy. And in other case it's not hard to find it again

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -75,20 +76,37 @@
 
+
+    /**
+     * Creates instance from {@link ObjectNode}
```
"more compliment" should be "more complient" in the commit name I think.

I think, if it's not too hard, that it would be nice to have it merged in the commit where you add/modify those method. 

What I'd do is:

I'll rebase interactively onto NF: Migrate from org.json.**** to jackson treemodel

Then I'll cherry pick this commit.

There will be conflict, because they can't do the correction on method that do not yet exists. That's great. You can delete all conflict and keep only the javadoc of method that exists at this step in the process.

Then you end the cherry pick and you end the interactive rebase.
In general, it works correctly and will detect that the cherry picked commit has already been partially applied, so it will only apply the part that were not applied. And you have easily split your commit in two

Arthur-Milchior replied:
Please ask question or even 1-1 time if there is any difficulty. I do realize that it's not necessarily trivial and I do not know whether there is any good tutorial for this technique or whether I devised it myself

TarekkMA replied:
I was going to merge this commit with "move to jackson tree model api" commit. since they are all part of the migration atomic step.

I will be sure to ping you if I have any difficulty

Arthur-Milchior reviewed (changes requested):
Really impressive work, and impressive time result too. 
I really appreciate the evolution in your atomic commit habit. It's really easy to read now. It's long, but there was no difficulty along the way. I already wrote it, so I won't repeat it on future PR, but I wanted to emphasize it one last time

A lot of comments. Feel free to disagree if I missed something and you think some of them are wrong and you had good reason. Sometime I avoided repeating the same comment many time, since there is a lot of unavoidable code repetition, at least I can avoid to repeat comment. 

Thanks a lot for this hard and useful work.

Arthur-Milchior commented:
Also, please correct the unit test failure

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:
```diff
@@ -144,12 +144,12 @@
             // review options
-            "'activeDecks': [1], " + "'curDeck': 1, " + "'newSpread': " + Consts.NEW_CARDS_DISTRIBUTE + ", "
-            + "'collapseTime': 1200, " + "'timeLim': 0, " + "'estTimes': True, " + "'dueCounts': True, "
+            "\"activeDecks\": [1], " + "\"curDeck\": 1, " + "\"newSpread\": " + Consts.NEW_CARDS_DISTRIBUTE + ", "
```
In android studio you can annotate variables with `@Language("JSON")` and the IDE will show syntex highlighting inside the string. Also, you can edit the string in an IDE window using the actions menu and it will remove `\` and make it nice to be edited.

![Screen Shot 2021-05-14 at 9 34 23 PM](https://user-images.githubusercontent.com/6633545/118320250-3e9a7400-b4fc-11eb-97ad-ec99148a2a72.png)

TarekkMA replied:
We can use files if we are inside pure unit test or robolectric test
```java
File file = new File("./src/test/assets/one_model.json");
```

If we are inside android test we can use
```java
context.getAssets().open("path")
```

Arthur-Milchior replied:
Cool. Do you accept to take care to add one of those?

TarekkMA replied:
Oh you mean production code not test code. The issue is we need `Context` reference and we will wait for the content to be read from file instead of having them as static strings.

I guess we could achieve it but will require some thinking of how to achieve it. 

I know that collection has `mContext` but I don't know if we should continue using `Context` object inside `Collection` class more.

TarekkMA commented on AnkiDroid/src/test/java/com/ichi2/utils/JSONObjectTest.java:
```diff
@@ -83,7 +82,6 @@ public void formatTest() {
         Assert.assertThrows(JSONException.class, () -> new JSONObject(extraOpeningBracket));
         Assert.assertThrows(JSONException.class, () -> new JSONObject(noClosingBracket));
         Assert.assertThrows(JSONException.class, () -> new JSONObject(wrongKeyValueSeparator));
-        Assert.assertThrows(JSONException.class, () -> new JSONObject(duplicateKey));
```
Jackson uses the last value and replaces existing ones.

I think we can configure jackson to fail on duplicates using this code
https://stackoverflow.com/a/37153217

jackson have lots of configurations but I didn't want to add ones unless we had a reason, for example I could configured Jackson to accept unquoted keys or be case insensitive

Arthur-Milchior replied:
I admit there is no reason to configure it I can think of. My remark is not about an actual need of AnkiDroid but having easy way to check what is changing

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -4021,17 +4021,15 @@ public String init(String jsonData) {
 
             try {
                 data = new JSONObject(jsonData);
-                if (!(data == JSONObject.NULL)) {
```
Can data be null anyways? I checked `org.json.JSONObject` and constructing a JSONObject can't result in JSONObject.

And yes that was strange the `equals` method is not overridden, so I overridden it in my version to be a reference comparison of the underlying jackson `JsonNode` to be compliant with the original behavior.

Arthur-Milchior replied:
No idea whether it is possible. I've never actually used JS addons. 
Anyway, it is the same issue before and after your PR, so I suggest we ignore it right now and wait until we see a reason to consider it if ever

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -42,94 +42,48 @@
+import static java.lang.System.identityHashCode;
 
-public class JSONArray extends org.json.JSONArray {
-    public JSONArray() {
```
I will try to achive that

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -42,94 +42,48 @@
-    public JSONArray() {
-        super();
-    }
+public class JSONArray {
```
I chose not to inherit from `ObjectNode` so I can control the API of the class.

What caused issues while I migrated to Jackson is API from `org.json.JSONObject` that is being used, I had to track all usages and make sure I got everything covered.

I will add this to the commit message

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
+        return this;
+    }
+
+    private JsonNode objectToJsonNode(Object value) {
```
yeah you are right I will try to refactor and use it in JSONArray and JSONObject

TarekkMA replied:
Added `JSONUtils` class with this method

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
         }
     }
 
     public JSONArray put(int index, boolean value) {
```
okay I will

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
-        }
+        put(index, objectToJsonNode(value));
```
I will see if I can get this implemented in jackson

Not at all, I'm here to learn so any kind of advice is always welcome.

Arthur-Milchior replied:
I admire the fact that you understood my message, while I wrote "javadoc" instead of "jackson".

TarekkMA replied:
I did a PR with it.
https://github.com/FasterXML/jackson-databind/pull/3156

Arthur-Milchior replied:
Nice. That's cool to give back to the libraries we use. 
I wonder how many thousands of dev will have used a method you wrote in one year

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
-        }
+        put(index, objectToJsonNode(value));
```
I changed `put(int index, Object value)` to be like this

```java
    public JSONArray put(int index, Object value) {
        JsonNode node = value instanceof JsonNode ? (JsonNode) value : objectToJsonNode(value);
        while (length() <= index) {
            put(JSONObject.NULL);
        }
        mNode.set(index, node);
        return this;
    }
```

Arthur-Milchior replied:
I'm not convinced it is an improvement.
In term of speed of the function, it does not really change anything. It just change slightly the order of the if branches.
In term of readability, it is a long line. Not an hard one, but I find it less readable than the last version

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
-            return this;
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
+        JsonNode node = objectToJsonNode(value);
```
good idea

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
-            throw new JSONException(e);
+        JsonNode node = objectToJsonNode(value);
+        while (length() <= index) {
+            put(JSONObject.NULL);
```
okay

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
+
+    protected void throwIfInvalidIndex(int index) {
+        if ((index < 0) || (index >= length())) {
+            throw new JSONException("Index:" + index + " is out of bounds");
```
Yes but I should wrap it with `JSONException` since I remember that some tests were failing because there was some production code that depended on catching it when adding to `JSONArray` object.

Arthur-Milchior replied:
I would be fine with having the test changed. Catching a more specific exception seems better. In this case, I find that `IndexOutOfBoundsException` is more specific about the error than `JSONException` since we already know you're dealing with JSON

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
         }
     }
 
     public boolean getBoolean(int index) {
```
okay

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -423,4 +363,60 @@ public Long next() {
     }
+
+
+    public boolean isNull(int i) {
```
okay

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -423,4 +363,60 @@ public Long next() {
+
+    @Override
+    public int hashCode() {
+        return identityHashCode(mNode);
```
`JsonNode` does override it to return the hashcode of the children. while `org.json.JSONObject` doesn't

Arthur-Milchior replied:
Thanks. I misunderstood first what this method does. I now understand, you're right.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -1,34 +1,34 @@
-/*  
+/*
```
done

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -1,34 +1,34 @@
-/*  
+/*
```
okay

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,101 +49,53 @@
-    public JSONObject(Map copyFrom) {
-        super(copyFrom);
+
+    public JSONObject(ObjectNode node) {
```
Okay I will add that,

I think it should be public because it's marked by `@JsonCreator` don't know if Jackson can access it if it's private/protected

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,101 +49,53 @@
+        try {
+            mNode = (ObjectNode) AnkiSerialization.getObjectMapper().readTree(source);
+        } catch (Exception e) {
+            throw new JSONException(e);
```
I do encapsulate exceptions because some part of the code depend on catching `JSONException`

Arthur-Milchior replied:
I would like it to be eventually changed. If you know what kind of exception is actually sent, it would be nice to just catch the jackson exception instead of adding an encapsulation level.
But you were right not to do it in this commit. It's better to ensure full compatibility.

TarekkMA replied:
I will try to handle the code that depends on catching `JSONException` in another PR.

Arthur-Milchior replied:
Thanks

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -153,199 +105,202 @@ public JSONObject(JSONObject copyFrom) {
+        return this;
     }
 
+
```
that's true, but would have exposed more API than we need to outside callers.

TarekkMA commented:
Thank you for your feed back I will work on it and will mention you when it's all done.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -154,153 +108,135 @@ public JSONArray(Collection<?> copyFrom) {
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
-        }
+    public ArrayNode getJsonNode() {
```
I need it so I can construct inheriting objects without deep copying the whole node

```java
    public Deck(JSONObject json) {
        super(json.getJsonNode());
    }
```

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -153,199 +105,202 @@ public JSONObject(JSONObject copyFrom) {
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
-        }
+    public JSONObject put(String name, CharSequence value) {
```
This one
https://github.com/ankidroid/Anki-Android/blob/ce42ab2d7cea7d060e873a2b3d374266685b9b6b/AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java#L422

TarekkMA commented:
@Arthur-Milchior I think it's ready for another round of review, I added some tests and made the implementation more compatible with the previous one.

Sorry for the long commit, but all items there seems atomic to me.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -50,88 +56,60 @@
-        super();
+import static java.lang.System.identityHashCode;
+
+public class JSONArray {
```
I'd like some javadoc, with at leat the following details.

An array of JSON elements. It follows `org.json.JSON*`  API since AnkiDroid used it previously. The only difference between that `getJSONArray` returns a com.ichi2.utils.JSONArray instead of a `org.json.JSONArray`, and same for `jsonObject` and `Null`.

 It uses Jackson ArrayNode internally. This should be invisible to the user as long as they do not use access to the underlying `mNode`. In particular, the `mNode`'s content is `JsonNode`, so it will not contain any `Double` nor any `com.ichi2.utils.JSONArray` but `DoubleNode` and `ArrayNode`.

Arthur-Milchior replied:
As last time, the same remark applies to JSONObject

TarekkMA replied:
okay

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -50,88 +56,60 @@
+    /**
+     * @return the backing {@link ArrayNode}
+     */
+    public ArrayNode getRootJsonNode() {
```
Can you use private? Or maybe protected so that only JSONObject use it.
If no, at least note in which case it's okay to use it.

TarekkMA replied:
I guess it could be protected

TarekkMA replied:
no I guess we need it public so we have the ability to convert `JSONObject` to one of it's sub classes

```java
public static Deck from(JSONObject json) { return new Deck(json.getRootJsonNode()) }

```

TarekkMA replied:
```java
    /**
     * @return the backing {@link ObjectNode}
     *
     * Changing the content of the returned node will result
     * in a change in this instance of JSONObject.
     *
     * This method should not be used directly to change the
     * content of the json object, but can be used for jackson
     * deserialization and for changing the wrapper class around
     * the {@link ObjectNode} like {@link com.ichi2.libanki.Deck#from(JSONObject)}
     */
    @NonNull
    @JsonValue
    public ObjectNode getRootJsonNode() {
        return mNode;
    }
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,303 @@
+    /**
+     * @return the backing {@link ObjectNode}
+     */
+    public ObjectNode getRootJsonNode() {
```
we might want it package private

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,303 @@
+    }
+
+    public JSONObject put(@NonNull String name, boolean value) {
+        return this.put(name, (Boolean) value);
```
`put(name, boolean.valueOf(value))`
as in array
to avoid a sequence of `instanceof`

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,303 @@
-            throw new JSONException(e);
+    public Object get(@NonNull String name) {
+        if (!has(name)) {
+            throw new JSONException("Key:" + name + " does not exist");
```
You might prefer having the same text in `get` and `getNode`
Actually, you may even avoid code duplication with:
```java
public Object get(@NonNull String name) {
  return JSONUtils.jsonNodeToObject(getNode(name));
}
```

TarekkMA replied:
get will now call getJsonNode

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Deck.java:
```diff
@@ -36,8 +36,7 @@ public Deck() {
     public Deck deepClone() {
-        Deck clone = new Deck();
-        return deepClonedInto(clone);
+        return new Deck(super.deepClone());
```
If you change the constructor as I asked, then please also change this method. Otherwise you'll get a useless double deep cloning. Same for DeckConfig and Model

TarekkMA replied:
Changed them to be:
```java
new Deck(this);
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -130,11 +130,7 @@ public JSONObject(JSONObject copyFrom) {
-    }
-
-    public JSONObject put(@NonNull String name, boolean value) {
+    public JSONObject put(String name, boolean value) {
```
Why did you remove `@NonNull` before name?

TarekkMA replied:
it must be a mistake will revert it

Arthur-Milchior reviewed (changes requested):
Thanks a lot for all of those changes. Really appreciate the improvement. Still some change requested. We are on the good path.

Sorry, there are things I could have asked previously, but I didn't really thought about it. It's easier to see the change I request now, since more obvious changes are already done.

I think almost everything is correct, it's almost only question of readability (feel free to disagree), and small performance gain (not sure of this, since we have not measured. But it seems not to hurt anyway to avoid useless sequences of `instanceOf`s)

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/Deck.java:
```diff
@@ -22,7 +22,7 @@
 public class Deck extends JSONObject {
     public Deck(JSONObject json) {
-        super(json);
+        super(json.getRootJsonNode());
```
Oh didn't think of the compatibility with the previous version.

I can make a static factory method
```java
    public Deck(JSONObject json) { super(json); }
    public Deck(ObjectNode node) { super(node); }
    public static Deck from(JSONObject json) { return new Deck(json.getRootJsonNode()) }
```

TarekkMA replied:
implemented it as follows

```java
    /**
     * @see  Deck#from(JSONObject)
     */
    @JsonCreator
    protected Deck(ObjectNode node) {
        super(node);
    }


    /**
     * Creates a Deck object from an {@link ObjectNode}
     *
     * NOTE: The passed node will be used directly, so
     * any change in the node will result in a change in
     * this object
     */
    public static Deck from(JSONObject json) {
        return new Deck(json.getRootJsonNode());
    }


    /**
     * Creates a copy from {@link JSONObject} and use it as a string
     *
     * This function will perform deepCopy on the passed object
     *
     * @see Deck#from(JSONObject) if you want to create a
     *                            Deck without deepCopy
     */
    public Deck(JSONObject json) {
        super(json);
    }
```

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
+     * Construct {@link JSONArray} filed with a given collection
+     */
+    public JSONArray(Collection<?> collection) {
+        this();
```
I resued Iterable constructor but made it nullable to accommodate the if the collection is null

TarekkMA replied:
```java
    public JSONArray(@Nullable Collection<?> collection) {
        this(collection == null ? null : collection.iterator());
    }
```

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
-            return this;
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
+        JsonNode node = JSONUtils.objectToJsonNode(value);
```
yes and put node version will be used to illuminate instance of checking

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
-            return super.getBoolean(index);
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
+        Object object = get(index);
```
Oh, I think `JSONObject` should be changed to use `JSONTypeConverters`, since Jackson does conversion different to `json.org`.

One example is boolean: `json.org` only convert string, while Jackson will convert lots of things to boolean like numbers if not zero will be true, etc.

using `json.org` is string value is "FaLSe" it will be used as `false` but in Jackson only lower case will be converted.

jackson should be faster since it doesn't rely on instanceof calls.

your proposed solution is okay but will not do any conversion like `json.org`

TarekkMA replied:
we can diverge from the json.org API once everything is stable with this implementation

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
         try {
-            return super.join(separator);
-        } catch (org.json.JSONException e) {
+            return (T) value;
```
you are right, the `ClassCastException` is thrown not when casting. Didn't know that.

TarekkMA replied:
used you snipped and it worked.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
+     * @return node at index
+     * @throws JSONException if the index is out of bounds
+     */
+    public JsonNode getJsonNode(int index) {
```
I think the only place it will be needed public is when wanting to convert `JSONObject/Array` to a subtype without copying

TarekkMA replied:
oh this is the same as `getRootJsonNode`, I should remove it

TarekkMA replied:
I think you are right, I got it mixed with `getRootJsonNode`, I will make it protected.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -50,88 +56,60 @@
+    /**
+     * Deserializes a string to to a {@link JSONArray} using the default {@link ObjectMapper}
+     * @param source the json string
+     */
```
done documented it.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
+    /**
+     * Construct {@link JSONArray} filed with items in the given iterator
+     */
+    public JSONArray(Iterator<?> iterator) {
```
@Nullable because it will be resued for the collection constructor

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
+     * Appends {@code value} to the end of this array.
+     */
+    public JSONArray put(int value) {
+        return put((Integer) value);
```
all methods but the `Object` one will add node directly.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +122,303 @@ public JSONArray(Object array) {
+     */
+    protected void throwIfInvalidIndex(int index) {
+        if (isInvalidIndex(index)) {
+            throw new JSONException(new IndexOutOfBoundsException("Index:" + index + " is out of bounds"));
```
changed the message,

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,303 @@
-            return this;
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
+    private JsonNode getNode(@NonNull String name) {
```
changed it to public

TarekkMA commented:
@Arthur-Milchior Thank you for the detailed feedback I think I have addressed all the issues, and will not directly insert node directly without the need for instance of checks.

Also added extra docs, and I'm now using `Optional` internal to handle `optXxx()` methods inside `JSONObject`

Can you please explain what is the issue with readability? Hoping once we make sure JSONObject/Array are stable, we can introduce changes to limit the API and make it simpler on other PRs. And I also think we shouldn't use `JSONObject/Array` directly all over the code base but rather the model objects them.

TarekkMA commented:
Opps, will look into the test errors

Arthur-Milchior commented:
I do not know what you mean about "readability"
Generally, this PR is huge. That make it hard to review thoroughly, so it is even more important that everything is as obvious as possible, reduce the number of line. I tried to improve it

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,377 @@
+
+    /**
+     * @return the value at "name" converted to double.
+     * if value cannot be converted, 0.0 is returned
```
Javadoc and implementation mismatch. I believe that if value can not be converted, then an exception is raised should be indicated here and in nearby functions

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,377 @@
+
+    /**
+     * @return array at "name"
+     */
```
For the first time, you do not indicate what occurs if value can't be converted. That's strange. You may want to correct that

TarekkMA replied:
fixed

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONTypeConverters.java:
```diff
@@ -0,0 +1,136 @@
+        }
+
+        public static JSONException typeMismatch(Object indexOrName, Object actual,
+                                                          String requiredType) throws JSONException {
```
Instead of a type, you can take a Class<?>.
You can take it with `int.class` for example for ints

Arthur-Milchior replied:
Same remark below

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +150,311 @@ public JSONArray(Object array) {
-            throw new JSONException(e);
+        Object object = get(index);
+        Boolean result = JSONTypeConverters.toBoolean(object);
+        if (result == null) {
```
You can probably win a lot of line if you added a static function

```java
protected static <T> T throwOnNull<T>(Object indexOrName, Object object, T result, String type) {
  if (result == null) {
    throw JSONTypeConverters.typeMismatch(index, object, type);
  }
return result;
}
```

TarekkMA replied:
good idea

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +150,311 @@ public JSONArray(Object array) {
+     */
+    protected <T> T getCasted(int index, Class<T> type) {
+        Object value = get(index);
+        if (value.getClass().equals(type)) {
```
You can probably factorize with a static method, used in both JSONObject and JSONArray
```java
protected static <T> getCasted(int index, Class<T> type, Object value) {
   if (value.getClass().equals(type)) {
     return (T) value;
  }
  throw JSONTypeConverters.typeMismatch(index, value, type.getName());
}
```
instead of duplicating code in both classes

TarekkMA replied:
I have added `JSONTypeConverters. convert/Or(....)` which will replace this
```java
// return JSONArray or null
return JSONTypeConverters.convertOr(value, JSONArray.class, null); 
// must return JSONObject or throw
return JSONTypeConverters.convert(name, get(name), JSONObject.class);
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,377 @@
     }
 
+    private Optional<Object> opt(@Nullable String name) {
+        return Optional.ofNullable(name).map(mNode::get).map(JSONUtils::jsonNodeToObject);
```
I do not see the point of this implementation.
I agree it seems correct. But I don't really get why you want to use this instead of if/else. It does not seem efficient since java is not created to deal efficiently with functionals and I do not find it easier to read either.
I also don't understand why you want Optional instead of Nullable. 
Can you please javadoc the reason
And also indicate that the Optional is `@NonNull`

TarekkMA replied:
Since it was an internal implementation not public, I thought I could use `Optional` and make use of it's `map` function which will only execute on nonnull.

The alternative is to check null after each step and return it if null.

It seemed like a cool idea to see what I can do with `Optional` 😄  but I can revert it back to the previous implementation using if/else.

TarekkMA replied:
Moved to normal `if`/`else`

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -363,18 +438,113 @@ public JSONObject deepClone() {
     }
+
+
+    public boolean optBoolean(@Nullable String name, boolean defaultValue) {
```
You could even avoid that much repetition by having a `optCasted` with a third argument with a default value, so that you don't need to repeat this line over and over again with a different type

TarekkMA replied:
`JSONTypeConverters. convert/Or(....)` which will factorize `optXxx()` calls and conversion calls to from boath `JSONArray` and `JSONObject`

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONTypeConverters.java:
```diff
@@ -0,0 +1,136 @@
+            return d;
+        }
+
+        public static Boolean toBoolean(Object value) {
```
I would like `@Nullable` everywhere. As it's important to know what occurs if the value is not a boolean. Since the javadoc of the class only mention the exception type, it would have been excepted that an exception is raised.

TarekkMA replied:
should be fixed

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Model.java:
```diff
@@ -40,14 +41,48 @@
+     * any change in the node will result in a change in
+     * this object
+     */
+    public static Model from(JSONObject json) {
```
Why do we need this method?

TarekkMA replied:
This is an alternative way to create Model from JSONObject without deepCopy.  We can also make the ObjectNode constructor public but I don't think we would use ObjectNode directly in the code

Arthur-Milchior replied:
I agree with the last point. We should avoid using ObjectNode directly. 
Thinking about it, it would also be a good thing to lint, ensuring that jackson is not directly used except where we expect it

Arthur-Milchior reviewed (commented):
I moved three commits as separate PR, so that the current PR get smaller.

Still a few change I'd like. really getting better, thanks a lot

TarekkMA commented:
Thank you for moving these to different PRs this is cleaner now.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +150,311 @@ public JSONArray(Object array) {
+    /**
+     * Construct {@link JSONArray} filed with a given collection
+     */
+    public JSONArray(@Nullable Collection<?> collection) {
```
Can you please explain this further?

TarekkMA replied:
Oh okay, I got what you mean, will do it

Arthur-Milchior commented:
Is it ready for re-review?

TarekkMA commented:
@Arthur-Milchior yes, it is. I was waiting to rereview your comments again before pining you.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +150,311 @@ public JSONArray(Object array) {
+    /**
+     * Appends {@code value} to the end of this array.
+     */
+    public JSONArray put(int value) {
```
tried to annotate everything

TarekkMA commented:
Done. rechecked/replied to all your comments

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONTypeConverters.java:
```diff
@@ -0,0 +1,233 @@
+     * value couldn't be converted
+     */
+    @Nullable
+    public static Boolean toBoolean(Object value) {
```
`@Nullable Object` (assuming we actually keep the `Object` type instead of a more specific one

TarekkMA replied:
Why keep it as object, this code is supposed to be the exact same class as in `org.json.JSON` internal class.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Deck.java:
```diff
@@ -16,19 +16,58 @@
+
+
+    /**
+     * Creates a Deck object from an {@link ObjectNode}
```
that's strange, since the method takes a JSONObject. You may want to state that it uses the underlying ObjectNode

TarekkMA replied:
right that is a mistake

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -79,6 +81,15 @@
+ *
+ * Databinding API:
+ *
+ * JSONObject support jackson databinding API, this is achieved @{@link JsonCreator} and
```
Do you want the two `@`?

TarekkMA replied:
no, this is a typo

Arthur-Milchior commented:
In case you didn't see, @TarekkMA, you know have plenty of conflict to merge. I'm kind of sorry for you, but they should be relatively easy fix since it does not seems there are many cases where you actually have to merge two different implementations.

I sincerely wonder how hard it will be to actually have the new test passes. Hopefully, not at all, but I would sincerely be surprised is everything went perfectly smoothly, at least because I had to add new methods as the one inherited failed the tests.

This is consistent with my previous suggestion, to generate a separate library just to "uses jackson speed and json API". Passing the same test should hopefully help ensuring the correctness of your implementation.

TarekkMA commented:
@Arthur-Milchior Thank you I will try to solve these issues. Passing these tests would be a big validation that nothing (hopefully) would break because of it.

TarekkMA commented:
I will try to address you comments first then do a rebase

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +149,285 @@ public JSONArray(Object array) {
+     * Appends {@code node} to the end of this array.
+     */
+    @NonNull
+    public JSONArray put(JsonNode node) {
```
Right that is cool, I wouldn't have thought about it like that if you didn't mention it.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +149,285 @@ public JSONArray(Object array) {
+     * This is used primarily to be compatible with the previous implementation
+     * which did wrap every exception inside {@link JSONException}
+     *
+     * @implNote note that jackson don't throw on out of bound exception
```
A quick look at jackson code reveals that only `set(index, JsonNode)` throw, but `get` will return null.

I added more info in the javadocs, what do you suggest the name should be ?

Arthur-Milchior replied:
The name is fine. I wanted the information in the javadov

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -144,163 +149,285 @@ public JSONArray(Object array) {
+        return new JSONArray(node);
     }
 
     public Iterable<JSONArray> jsonArrayIterable() {
```
yeah, that is right.

 the one returning null on empty is somewhat unintuitive why we do that.

Arthur-Milchior replied:
No idea

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONArray.java:
```diff
@@ -423,4 +554,38 @@ public Long next() {
+        return mNode.toString();
+    }
+
+    public String toPrettyString() {
```
no documentation about this method, but I guess it will format the object as a formatted json.
I think the only use is debugging.

we can remove it if we don't need it right now.

but note that the original implementation have `String toString(int indentSpaces)`, and this may be a substitute for taht

Arthur-Milchior replied:
Actually, in JsonNode.java, you can read:

>      /**
>      * Method that will produce (as of Jackson 2.10) valid JSON using
>      * default settings of databind, as String.
>      * If you want other kinds of JSON output (or output formatted using one of
>      * other Jackson-supported data formats) make sure to use
>      * {@link ObjectMapper} or {@link ObjectWriter} to serialize an
>      * instance, for example:
>      *<pre>
>      *   String json = objectMapper.writeValueAsString(rootNode);
>      *</pre>
>      *<p>
>      * Note: method defined as abstract to ensure all implementation
>      * classes explicitly implement method, instead of relying
>      * on {@link Object#toString()} definition.
>      */
>     @Override
>     public abstract String toString();
> 
>     /**
>      * Alternative to {@link #toString} that will serialize this node using
>      * Jackson default pretty-printer.
>      *
>      * @since 2.10
>      */
>     public String toPrettyString() {
>         return toString();
>     }

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,369 @@
+     * Creates an empty json object using the default {@link ObjectMapper}
+     */
+    public JSONObject() {
+        this(AnkiSerialization.getObjectMapper().createObjectNode());
```
`ObjectNode`/`ArrayNode` require `JsonNodeFactory` to be created. So instead of hardcoding the `JsonNodeFactory` config about how to handle decimals, I use the globally created `ObjectMapper` and it will create object/array with the correct configured factory.

```java
// copied from Jackson code
    public static JsonNodeFactory withExactBigDecimals(boolean bigDecimalExact)
    {
        return bigDecimalExact ? decimalsAsIs : decimalsNormalized;
    }
```

TarekkMA replied:
I use a getter style rather than a public field so if we need to have more in-depth configurations we can do it in the method. and create it lazily.

If you see that as not very important we can convert to a public field.

Arthur-Milchior replied:
Nice
I just want all of this in code comment, please

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONObject.java:
```diff
@@ -49,303 +50,369 @@
-        } catch (org.json.JSONException e) {
-            throw new JSONException(e);
+    @NonNull
+    public JsonNode getJsonNode(@NonNull String name) {
```
no this one can be private, I also confused it with `getRootJsonNode` sorry

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONTypeConverters.java:
```diff
@@ -0,0 +1,233 @@
+     * if it couldn't be converted
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> T convertOr(@NonNull Object value, @NonNull Class<T> type, T defaultValue) {
```
I don't think I get why you want to make `value` as `JsonNode`?

The goal is to convert different values to the target type.
String -> Boolean
Integer - > Boolean
etc...
<hr>
If I have `JsonNode` I will have to handle each type first (`BoolNode`, `DoubleNode`, `IntNode`, etc), and after that I will try to convert the acquired value to the target type.
And this is exactly what I'm doing right now.

For example in `optBoolean` I try to get `JsonNode` at `name` and then convert that node to a value using `JSONUtils.jsonNodeToObject(node)`, and then convert the returned value using `JSONTypeConverters.convertOr`

```java
    public Object opt(@Nullable String name) {
        if (name == null) {
            return null;
        }
        JsonNode node = mNode.get(name);
        if (node == null) {
            return null;
        }
        return JSONUtils.jsonNodeToObject(node);
    }
    public boolean optBoolean(@Nullable String name, boolean defaultValue) {
        Object value = opt(name);
        if (value == null) {
            return defaultValue;
        }
        return JSONTypeConverters.convertOr(value, Boolean.class, defaultValue);
    }
```

Arthur-Milchior replied:
My bad, I misunderstood what was done here

Arthur-Milchior replied:
I would appreciate a top level javadoc comment where you indicate that `Object` value must be one of String, Boolean, Integer, Double... as sadly it's too vague right now

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/JSONTypeConverters.java:
```diff
@@ -0,0 +1,233 @@
+
+    @NonNull
+    public static JSONException typeMismatch(Object indexOrName, Object actual,
+                                             String requiredType) throws JSONException {
```
done

Arthur-Milchior commented:
You may want to take a look at https://github.com/Arthur-Milchior/Anki-Android/commit/e4587b073a4f6d1b7d24b62759ff5c02a81119d8 and https://github.com/Arthur-Milchior/Anki-Android/commit/9d948beb1374e37421d062716defd890e6383a4c and https://github.com/Arthur-Milchior/Anki-Android/commit/6a6b2471d01b9b99e94cb9501c46fb1287aafc79 where I try to factorize your code

TarekkMA commented:
Thank you for the refactor, I think `JSONContainer` will defiantly remove the duplicate code and make it more maintainable.

Unfortunately, I will be not available for the next couple of weeks due to my final exams.

Arthur-Milchior commented:
I assume you mean "definitively" and not "defiantly" even if it seems to be an existing word. 

While it avoid duplicate code, it forces using `Integer` instead of `int` which will potentially lead to lost time doing conversions between type. So I don't suggest merging my proposed changes without first measuring the time it takes. Since the point of this migration is really to get json treated fastly, it would be a shame And my experience shows that `int`->`Integer` is expensive in for loops.

JSON is not a priority at all, it can clearly waits for your exams to end, there is no way it ever blocks anything. Thank you for the warning, and it'll easily wait for your come back. Good luck with those exams

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/utils/JSONTypeConvertersTest.java:
```diff
@@ -0,0 +1,131 @@
+                    {23, Integer.class, 23},
+                    {232.2D, Integer.class, 232},
+                    {"232", Integer.class, 232},
+                    {"22.2", Integer.class, 22},
```
can you check rounding here?

TarekkMA replied:
yes I will check that

david-allison-1 reviewed (commented):
IMO needs 3 reviewers as mine doesn't hold weight. I haven't used Jackson and am not familiar with the API.

I'd like to add a quick "full startup" perf statistic to the OP to prove this is worth the risk. I'm sure it will be

I'd like to see a little more testing about the difference between `NULL` and `null`, but other than that, no guidance

We have a relatively long time until 2.16, so we should give this a long soak in alpha.

Arthur-Milchior commented:
I hope the exams went well/are going well.
It seems you'll get plenty of conflict to resolve when you can come back

TarekkMA commented:
Exams are going okay, thank you.

I hope those will be easy to resolve once I get back.
